File: markov-chords/markov-chords.py
from time import sleep
import json
import random
import sys

sleep(1)

# weights = {
#   "cm": {"cm": 0.3, "f": 0.2, "fm": 0.2, "gm": 0.2, "ab": 0.1},
#   "fm": {"cm": 0.1, "fm": 0.1, "gm": 0.4, "ab": 0.2, "d": 0.2},
#   "f": {"cm": 0.1, "fm": 0.1, "gm": 0.4, "ab": 0.2, "d": 0.2},
#   "gm": {"cm": 0.3, "g": 0.3, "gm": 0.1, "fm": 0.2, "ab": 0.1},
#   "g": {"cm": 1},
#   "ab": {"gm": 0.5, "g": 0.4, "fm": 0.1},
#   "d": {"fm": 0.2, "gm": 0.5, "g": 0.3}
# }

weights = json.load(open(sys.argv[1]))

def choose_weighted(dic: dict[str, float])-> str:
    r = random.random()
    cum = 0
    for k, v in dic.items():
        cum += v
        if r < cum:
            return k

num_resolutions = 5
if '-m' in sys.argv:
    final_chord = 'cm'
else:
    final_chord = 'c'
curr_chord = final_chord

print("wide")
print("down")
print("tempo 60")
while True:
    print(curr_chord)
    print(curr_chord, file=sys.stderr)
    print("bsleep 4")
    last_chord = curr_chord
    curr_chord = choose_weighted(weights[curr_chord])
    if num_resolutions == 0:
        break
    if curr_chord == final_chord and last_chord != final_chord:
        num_resolutions -= 1

print("bsleep 4")
print("quit")

File: markov-chords/markov-chords.rb
require 'time'

sleep 0.1
chords = ['c', 'f', 'g', 'g7', 'c']

weights = {
  "c": {"c": 0.3, "f": 0.4, "g": 0.2, "am": 0.1},
  "f": {"c": 0.1, "f": 0.1, "g": 0.4, "am": 0.4},
  "g": {"c": 0.5, "g": 0.1, "g7": 0.3, "am": 0.1},
  "am": {"g": 0.6, "f": 0.4},
  "g7": {"g7": 0.1, "c": 0.9}
}


def choose_weighted map
  r = rand()
  cum = 0
  map.each_pair {|k, v|
    cum += v
    if r <= cum
      return k
    end
  }
end

# chords.each {|x| 
#   puts x
#   puts 'bsleep 4'
# }

puts weights[:c]
puts weights["c"]

curr_chord = "c"
10.times {|i|
  puts curr_chord
  puts 'bsleep 4'
  curr_chord = choose_weighted(weights[curr_chord])
}
puts 'quit'
